packages/runtime-common/realm.ts

```diff
@@ -666,37 +667,79 @@ export class Realm {
         `No challenge previously issued for user ${user}`,
       );
     }
-    let messages = await this.#matrixClient.roomMessages(roomId);
-    let latestChallenge = messages.find(
-      (m) =>
-        m.type === 'm.room.message' &&
-        m.sender === this.#matrixClient.userId &&
-        m.content.body.startsWith('auth-challenge:'),
-    );
-    if (!latestChallenge) {
-      return badRequest(
-        this,
-        `No challenge previously issued for user ${user}`,
-      );
+
+    // The messages look like this:
+    // --- Matrix Room Messages ---:
+    // realm1
+    // auth-challenge: 7cb8f904a2a53d256687c2aeb374a686a26cfd66af5fcc09a366d49644a3e2ba
+    // realm2
+    // auth-response: 342c5854-e716-4bda-9b31-eba83d24e25d
+    // ----------------------------
+
+    // This is a best-effort type of implementation - we don't know when the messages will appear in the room so we just wait for a bit.
+    // This is not a problem when the realms are on the same matrix server but when they are on different (federated) servers the latencies and
+    // race conditions can cause delays in the messages appearing in the room.
+    let waitForAuthMessages = async (
```

Comment:
consider separating a "waitForMessage" util from the details about the auth-challenge format. They are separate concerns and such a util might be independently useful.

Reply:
Agree, I refactored this